Library/Homebrew/extend/cached.rb

```diff
@@ -0,0 +1,49 @@
+# typed: strict
+# frozen_string_literal: true
+
+module Cached
+  module Clear
```

Comment:
> Why is a new module needed now?

To make the code more readable and easier to maintain, i.e. not having to manually keep track of which cached variables need to be cleared.

> Is this only used once?

So far only in `Tap`, but it should be possible to replace `Cachable` with this.

Reply:
I think it'd be good to use this in 2+ places if it's added here to verify the API is sufficiently generic and better test the edges.

Reply:
> So far only in `Tap`, but it should be possible to replace `Cachable` with this.

I think having both `Cachable` and `Cached` is a bit confusing especially if they are used in the same module. The long-term goal should be to only have one. I prefer the simplicity of `Cachable` in most cases. `Cached` is not as flexible and that means once we try to cache methods that take arguments the interface is not as clean and it's possible that we might need to add helper methods which actually add complexity.

I like the approach and I think the code is cool but I'm not sure it's worth the cost here from a maintainability point of view.

As a quick comparison of `Cachable` and `Cached`, I'll show some examples of things that would be easier and things that would be harder.

```rb
# Now (on master)
def git_head
  raise TapUnavailableError, name unless installed?

  @git_head ||= git_repo.head_ref
end

# Cached
cached def git_head
  raise TapUnavailableError, name unless installed?

  git_repo.head_ref
end

# Cachable
def git_head
  raise TapUnavailableError, name unless installed?

  cache[:git_head] ||= git_repo.head_ref
end
```

These functions are all similar complexity. The advantage of caching is being able to clear the cache with one easy method which would be provided by both `Cached` and `Cachable`.

Note: `Cached` and `Cachable` are not exactly equivalent here. `Cached` will only be evaluated once while `Cachable` will check if the tap is installed each time though the head ref will be cached.

Here's another example.

```rb
# Now (on master)
def audit_exceptions
  @audit_exceptions ||= begin
    ensure_installed!
    super
  end
end

# Cached
cached def audit_exceptions
  ensure_installed!
  super
end

# Cachable
def audit_exceptions
  cache[:audit_exceptions] ||= begin
    ensure_installed!
    super
  end
end
```

The complexity is not much different in this case either. I would argue that it's also easier to understand at a glance since it uses basic Ruby patterns as well but that's very subjective.

One fine example, that's a bit different.

```rb
# Cachable (on master)
def self.fetch(user, repo)
  ...

  cache_key = "#{user}/#{repo}".downcase
  cache.fetch(cache_key) { |key| cache[key] = Tap.new(user, repo) }
end

# Cached
def self.fetch(user, repo)
  ...

  _fetch(user, repo)
end

private_class_method cached_class_method def self._fetch(user, repo)
  new(user, repo)
end
```

The `Tap.fetch` method shows that `Cached` is a not as flexible and in this case that means we need to create an additional method. One method is not the end of the world, of course, but it does suggest that this doesn't always simplify the code.

I feel like the simplest approach would just be to use `Cachable` without any modifications.

Reply:
> The long-term goal should be to only have one.

I agree with most of the above except: I don't think this should be "long-term", I would really like to avoid having both existing separately for any time/at all.